xmldumps-backup/mwbzutils/docs/dumplastbz2block.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.40.4.
.TH DUMPLASTBZ2BLOCK "1" "July 2013" "dumplastbz2block 0.0.4" "User Commands"
.SH NAME
dumplastbz2block \- Find last bz2 block in bzip2 file and dump contents
.SH SYNOPSIS
.B dumplastbz2block
[\fI--version|--help\fR]
.br
.B dumplastbz2block
\fI<infile>\fR
.SH DESCRIPTION
Find the last bz2 block marker in a file and dump whatever can be
decompressed after that point.  The header of the file must be intact
in order for any output to be produced.
This will produce output for truncated files as well, as long as there
is 'enough' data after the block marker.
Exits with 0 if some decompressed data was written, 1 if no data could
be uncompressed and \fB\-1\fR on error.
.SH OPTIONS

Flags:
.TP
\fB\-h\fR, \fB\-\-help\fR
Show this help message
.TP
\fB\-v\fR, \fB\-\-version\fR
Display the version of this program and exit
.PP
Arguments:
.TP
<infile>
Name of the file to process
.SH AUTHOR
Written by Ariel T. Glenn.
.SH "REPORTING BUGS"
Report bugs in dumplastbz2block to <https://bugzilla.wikimedia.org/>.
.PP
.br
See also checkforbz2footer(1), dumpbz2filefromoffset(1), findpageidinbz2xml(1),
recompressxml(1), writeuptopageid(1)
.SH COPYRIGHT
Copyright \(co 2011, 2012, 2013 Ariel T. Glenn.  All rights reserved.
.PP
This program is free software: you can redistribute it and/or modify it
under the  terms of the GNU General Public License as published by the
Free Software Foundation, either version 2 of the License, or (at your
option) any later version.
.PP
This  program  is  distributed  in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
Public License for more details.
.PP
You should have received a copy of the GNU General Public License along
with this program.  If not, see <http://www.gnu.org/licenses/>


xmlfileutils/CHANGELOG
29 May 2013 Kent L. Miller <wpmirrordev@gmail.com>

   Release 0.0.2-1

    * mwxml2sql.c:
      Doc: show_version() rewritten for compatibility with help2man.
      Doc: usage() rewritten for compatibility with help2man.
      Doc: man page generated with help2man
    * sql2txt.c:
      Bug fixed: long options, --sqlfile and --textfile, did not work.
      Doc: show_version() written for compatibility with help2man.
      Doc: usage() rewritten for compatibility with help2man.
      Doc: man page generated with help2man
      Option: added -V, --version option for use with help2man
    * sqlfilter.c:
      Doc: show_version() written for compatibility with help2man.
      Doc: usage() rewritten for compatibility with help2man.
      Doc: man page generated with help2man
      Option: created -w, --version option for use by help2man
    * Makefile:
      Debian: build and deinstall targets written; clean target improved
      Debian: install target rewritten to use: install $(DESTDIR)/...
      Debian: dist renamed `mediawiki-mwxml2sql-0.0.2.tar.gz'
      Doc: man pages generated using `help2man'
      MWXML2SQLVERS for Makefile dist version and for VERSION in source
    * mwxml2sql.h:
      VERSION passed from Makefile

22 May 2013 Ariel T. Glenn <ariel@wikimedia.org>

    * mwxmlelts.c:
      handle text with <deleted /> instead of content
      workaround for XML files that may contain duplicate text ids

21 May 2013 Ariel T. Glenn <ariel@wikimedia.org>

    * mwxmlelts.c:
      convert revision timestamp from datestring format to db format

10 Apr 2013 Ariel T. Glenn <ariel@wikimedia.org>

    Release 0.0.2

    * sqlfilter: filters sql table dumps against e.g. list of pageids
      (and the column number of the pageid in the sql row) and writes
      only those tuples that match
    * wikiretriever.py: retrieves page titles and or content from a wiki
    * wikicontent2sql.py: downloads all pages containing a given template,
      swapping out talk pages for their corresponding pages (eg File talk
      for File etc.), converts them to page, revision and table sql,
      and filters downloaded sql table dumps according to the page ids
      of the downloaded content, producing a pile of sql files for import
    * package: various bugfixes

25 Jan 2013 Ariel T. Glenn <ariel@wikimedia.org>

   Release 0.0.1

   Features:
    *  Reads stub and text content xml files in parallel, producing separate
       sql files for inserts to page, revision and text tables
    *  Reads bz2, gz or plain text content, writes bz2, gz or plaintext output
       based on extensions of stub/page content files and output filename
    *  Reads xml dumps with schema 0.5 through 0.8, writes for MW 1.5 through
       1.20 (tested only for 0.7 in and MW 1.20 out)


xmldumps-backup/mwbzutils/dumplastbz2block.c
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <stdlib.h>
#include <errno.h>
#include <inttypes.h>
#include <getopt.h>
#include "mwbzutils.h"

void usage(char *message) {
  char * help =
"Usage: dumplastbz2block [--version|--help]\n"
"   or: dumplastbz2block <infile>\n\n"
"Find the last bz2 block marker in a file and dump whatever can be\n"
"decompressed after that point.  The header of the file must be intact\n"
"in order for any output to be produced.\n"
"This will produce output for truncated files as well, as long as there\n"
"is 'enough' data after the block marker.\n"
"Exits with 0 if some decompressed data was written, 1 if no data could\n"
"be uncompressed and -1 on error.\n\n"
"Options:\n\n"
"Flags:\n\n"
"  -h, --help       Show this help message\n"
"  -v, --version    Display the version of this program and exit\n\n"
"Arguments:\n\n"
"  <infile>         Name of the file to process\n\n"
"Report bugs in dumplastbz2block to <https://bugzilla.wikimedia.org/>.\n\n"
"See also checkforbz2footer(1), dumpbz2filefromoffset(1), findpageidinbz2xml(1),\n"
"recompressxml(1), writeuptopageid(1)\n\n";
  if (message) {
    fprintf(stderr,"%s\n\n",message);
  }
  fprintf(stderr,"%s",help);
  exit(-1);
}

void show_version(char *version_string) {
  char * copyright =
"Copyright (C) 2011, 2012, 2013 Ariel T. Glenn.  All rights reserved.\n\n"
"This program is free software: you can redistribute it and/or modify it\n"
"under the  terms of the GNU General Public License as published by the\n"
"Free Software Foundation, either version 2 of the License, or (at your\n"
"option) any later version.\n\n"
"This  program  is  distributed  in the hope that it will be useful, but\n"
"WITHOUT ANY WARRANTY; without even the implied warranty of \n"
"MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General\n"
"Public License for more details.\n\n"
"You should have received a copy of the GNU General Public License along\n"
"with this program.  If not, see <http://www.gnu.org/licenses/>\n\n"
    "Written by Ariel T. Glenn.\n";
  fprintf(stderr,"dumplastbz2block %s\n", version_string);
  fprintf(stderr,"%s",copyright);
  exit(-1);
}


int main(int argc, char **argv) {

  bz_info_t bfile;

  int fin;
  int result;
  buf_info_t *b;

  int firstblock = 1;
  int length = 5000; /* output buffer size */

  int optc;
  int optindex=0;

  struct option optvalues[] = {
    {"help", 0, 0, 'h'},
    {"version", 0, 0, 'v'},
    {NULL, 0, NULL, 0}
  };

  if (argc != 2) {
    usage("Missing option or argument.");
    exit(-1);
  }

  while (1) {
    optc=getopt_long_only(argc,argv,"hv", optvalues, &optindex);
    if (optc=='h')
      usage(NULL);
    else if (optc=='v')
      show_version(VERSION);
    else if (optc==-1) break;
    else usage("Unknown option or other error\n");
  }

  if (optind >= argc) {
    usage("Missing filename argument.");
  }

  fin = open (argv[optind], O_RDONLY);
  if (fin < 0) {
    fprintf(stderr,"failed to open file %s for read\n", argv[optind]);
    exit(-1);
  }

  bfile.file_size = get_file_size(fin);
  bfile.footer = init_footer();
  result = check_file_for_footer(fin, &bfile);
  if (result == -1) {
    bfile.position = bfile.file_size;
  }
  else {
    bfile.position = bfile.file_size - (off_t)11; /* size of footer, perhaps with 1 byte extra */
  }
  bfile.position -=(off_t)6; /* size of marker */
  bfile.initialized = 0;
  b = init_buffer(length);
  bfile.bytes_read = 0;

  /*  init_bz2_file(&bfile, fin, BACKWARD); */
  firstblock = 1;

  if (find_first_bz2_block_from_offset(&bfile, fin, bfile.position, BACKWARD) <= (off_t)0) {
    fprintf(stderr,"failed to find block in bz2file\n");
    exit(-1);
  }
  while ((get_buffer_of_uncompressed_data(b, fin, &bfile, FORWARD)>=0) && (! bfile.eof) && (! bfile.position == (off_t)0)) {
    if (bfile.bytes_read) {
      fwrite(b->next_to_read,b->bytes_avail,1,stdout);
      b->next_to_read = b->end;
      b->bytes_avail = 0;
      b->next_to_fill = b->buffer; /* empty */
      bfile.strm.next_out = (char *)b->next_to_fill;
      bfile.strm.avail_out = b->end - b->next_to_fill;
      firstblock = 0;
    }
    else {
      /* should never happen */
      fprintf(stderr,"there was a block but now it's gone, giving up\n");
      exit(-1);
    }
  }
  if (b->bytes_avail) {
    fwrite(b->next_to_read,b->bytes_avail,1,stdout);
    b->next_to_read = b->end;
    b->bytes_avail = 0;
    b->next_to_fill = b->buffer; /* empty */
    bfile.strm.next_out = (char *)b->next_to_fill;
    bfile.strm.avail_out = b->end - b->next_to_fill;
  }  
  close(fin);
  exit(0);
}
